llama-index-core/llama_index/core/utilities/token_counting.py

```diff
@@ -58,25 +59,33 @@ def estimate_tokens_in_messages(self, messages: List[ChatMessage]) -> int:
 
                 tokens += 3  # Additional tokens for function call
 
+            if "tool_calls" in additional_kwargs:
+                for tool_call in additional_kwargs["tool_calls"]:
+                    if (
+                        hasattr(tool_call, "function")
+                        and tool_call.function is not None
+                    ):
+                        tokens += self.get_string_tokens(tool_call.function.name)
+                        tokens += self.get_string_tokens(tool_call.function.arguments)
+
+                        tokens += 3  # Additional tokens for tool call
+
             tokens += 3  # Add three per message
 
-            if message.role == MessageRole.FUNCTION:
+            if message.role == MessageRole.FUNCTION or message.role == MessageRole.TOOL:
                 tokens -= 2  # Subtract 2 if role is "function"
 
         return tokens
 
-    def estimate_tokens_in_functions(self, functions: List[Dict[str, Any]]) -> int:
-        """Estimate token count for the functions.
+    def estimate_tokens_in_tools(self, tools: List[Dict[str, Any]]) -> int:
+        """Estimate token count for the tools.
 
-        We take here a list of functions created using the `to_openai_spec` function (or similar).
+        We take here a list of tools created using the `to_openai_tool()` function (or similar).
 
         Args:
-            function (list[Dict[str, Any]]): The functions to estimate the token count for.
+            tools (list[Dict[str, Any]]): The tools to estimate the token count for.
 
         Returns:
             int: The estimated token count.
         """
-        prompt_definition = str(functions)
-        tokens = self.get_string_tokens(prompt_definition)
-        tokens += 9  # Additional tokens for function definition
-        return tokens
+        return self.get_string_tokens(str(tools))
```

Comment:
1. tools in this case is FunctionTool not Dict[str, Any] right? 
2. why is str(tools) used to estimate token count

Reply:
It is, I pass in `[x.metadata.to_openai_tool() for x in tools]`

There isn't really a reliable way I could find online to count these tokens? This seems give a generous estimate compared to the usage reported by openai